float8_experimental/float8_dynamic_linear.py

```diff
@@ -96,8 +107,108 @@ def from_float(cls, mod, emulate: bool = False) -> "Float8DynamicLinear":
                 "bias": False,
             }
             new_mod = cls(**super_kwargs)
-        new_mod.weight = mod.weight
-        new_mod.bias = mod.bias
         new_mod.forward_config = ScaledMMConfig(emulate, True if not emulate else False)
         new_mod.backward_config = ScaledMMConfig(emulate, False)
+        if config.enable_fsdp_fp8_all_gather:
+            new_mod.weight = nn.Parameter(
+                WeightWithDynamicFloat8CastTensor(mod.weight, new_mod.forward_config)
+            )
+        else:
+            new_mod.weight = mod.weight
+        new_mod.bias = mod.bias
         return new_mod
+
+
+# FSDP pads its local tensor on dim-0. The subclass should be preserved such
+# that the padded local tensor (and any transformations like copying to GPU)
+# is of the subclass as well.
+_ops_to_preserve_subclass = {
```

Comment:
is there anything we need to support other than these ops and creating a float8tensor? If not, can we just support what we need and not the generic case?

Reply:
I think the user could do arbitrary ops on the parameters (e.g. gradient norm clipping, metrics calculations, EMA, etc.). For those, we do not need to preserve the subclass, but we should still consider the ops valid for the tensor.

I was thinking that it might be difficult to nail down all of the ops that we want to support. It seems okay to me to know which ops _should_ preserve the subclass since those would be the ones needed for sharding.